oid4vci/src/errors.rs

```
@@ -0,0 +1,123 @@
+pub struct OID4VCError<T>
+where
+    T: ResponseErrorType,
+{
+    pub error: T,
+    pub error_description: Option<String>,
+}
+
+impl<T> OID4VCError<T>
+where
+    T: ResponseErrorType,
+{
+    pub fn new(error: T) -> Self {
+        Self {
+            error,
+            error_description: None,
+        }
+    }
+
+    pub fn new_with_description(self, error_description: &str) -> Self {
+        Self {
+            error_description: Some(error_description.to_string()),
+            ..self
+        }
+    }
+}
+
+impl<T> Display for OID4VCError<T>
+where
+    T: ResponseErrorType + std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Error: {:?}", self.error)?;
+        if let Some(desc) = &self.error_description {
+            write!(f, " - {}", desc)?;
+        }
+        Ok(())
+    }
+}
+
+// - - - Notification Error Type !
+
+#[derive(Debug, Serialize, Deserialize, AsRefStr)]
+#[serde(rename_all = "snake_case")]
+pub enum NotificationRequestErrorResponse {
+    InvalidNotificationRequest,
+    InvalidNotificationId,
+    MissingNotificationParameter,
+    InvalidToken,
+}
+
+impl ResponseErrorType for NotificationRequestErrorResponse {}
+
+impl ErrorStatusCode for NotificationRequestErrorResponse {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            Self::InvalidNotificationRequest => StatusCode::BAD_REQUEST,
+            Self::InvalidNotificationId => StatusCode::BAD_REQUEST,
+            Self::MissingNotificationParameter => StatusCode::BAD_REQUEST,
+            Self::InvalidToken => StatusCode::UNAUTHORIZED,
+        }
+    }
+}
+
+// - - - Credential Error Type !
+
+#[derive(Debug, Serialize, Deserialize, AsRefStr)]
+#[serde(rename_all = "snake_case")]
+pub enum CredentialRequestErrorResponse {
+    InvalidCredentialRequest,
+    UnsupportedCredentialType,
+    UnsupportedCredentialFormat,
+    InvalidProof,
+    InvalidEncryptionParameters,
```

Looks like the `invalid_request` variant is missing referred to here: https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0-13.html#authorization-errors

---

I think it means in the spec that these error codes must be used _instead_ of invalid_request

---

That applies only for errors specific to the payload (request body) of the Credential Request. HTTP requests and responses consist typically of three parts: the Status Code, the Headers, and optionally a Body.

The Access Token is part of the Headers, so any error related to the Access Code should throw an `invalid_request` error response. For any error related to the request body it should throw one of the Credential Request Errors defined in the VCI spec :)

---

Ok so I went through all the Errors you've implemented now:

* `AuthorizationErrorResponse` --> as defined in RFC6749
* `TokenErrorResponse` --> as defined in RFC6749 (and with added clarifications in VCI)
* `CredentialErrorResponse` --> as defined in VCI,
  * Error Response related to Access Token as defined in RFC6750 (looks like should then be `invalid_token`)
  * `invalid_request` indeed should not be used because the Credential Error Responses indeed replace it.
* `BatchCredentialErrorResponse` ^idem
* `DeferredCredentialErrorResponse` ^idem (but with two extra error variants)
* `NotificationErrorResponse` --> as defined in VCI
  * Like the `CredentialErrorResponse`: Error Response related to Access Token as defined in RFC6750 (looks like should then be `invalid_token`)

I think this is a correct overview right? So then that would mean that `CredentialErrorResponse`, `BatchCredentialErrorResponse`, `DeferredCredentialErrorResponse` and `NotificationErrorResponse` still need the `invalid_token` (and not the `invalid_request` variant that I suggested earler)?